packages/eslint-plugin/src/rules/no-void-expression.ts

```diff
@@ -0,0 +1,286 @@
+      invalidVoidExpr:
+        'Placing a void expression inside another expression is forbidden. ' +
+        'Move it to its own statement instead.',
+      invalidVoidExprWrapVoid:
+        'Void expressions used inside another expression ' +
+        'must be moved to its own statement ' +
+        'or marked explicitly with the `void` operator.',
+      invalidVoidExprArrow:
+        'Returning a void expression from an arrow function shorthand is forbidden. ' +
+        'Please add braces to the arrow function.',
+      invalidVoidExprArrowWrapVoid:
+        'Void expressions returned from an arrow function shorthand ' +
+        'must be marked explicitly with the `void` operator.',
+      invalidVoidExprReturn:
+        'Returning a void expression from a function is forbidden. ' +
+        'Please move it before the `return` statement.',
+      invalidVoidExprReturnLast:
+        'Returning a void expression from a function is forbidden. ' +
+        'Please remove the `return` statement.',
+      invalidVoidExprReturnWrapVoid:
+        'Void expressions returned from a function ' +
+        'must be marked explicitly with the `void` operator.',
+      voidExprWrapVoid: 'Mark with an explicit `void` operator',
+    },
+    schema: [
+      {
+        type: 'object',
+        properties: {
+          ignoreArrowShorthand: { type: 'boolean' },
+          ignoreVoidOperator: { type: 'boolean' },
+        },
+        additionalProperties: false,
+      },
+    ],
+    type: 'problem',
+    fixable: 'code',
+  },
+  defaultOptions: [{}],
+
+  create(context, [options]) {
+    return {
+      'AwaitExpression, CallExpression, TaggedTemplateExpression'(
+        node:
+          | TSESTree.AwaitExpression
+          | TSESTree.CallExpression
+          | TSESTree.TaggedTemplateExpression,
+      ): void {
+        const parserServices = util.getParserServices(context);
+        const checker = parserServices.program.getTypeChecker();
+        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(node);
+        const type = util.getConstrainedTypeAtLocation(checker, tsNode);
+        if (!tsutils.isTypeFlagSet(type, ts.TypeFlags.VoidLike)) {
+          // not a void expression
+          return;
+        }
+
+        const invalidAncestor = findInvalidAncestor(node);
+        if (invalidAncestor == null) {
+          // void expression is in valid position
+          return;
+        }
+
+        const sourceCode = context.getSourceCode();
+        const wrapVoidFix = (fixer: TSESLint.RuleFixer): TSESLint.RuleFix => {
+          const nodeText = sourceCode.getText(node);
+          const newNodeText = `void ${nodeText}`;
+          return fixer.replaceText(node, newNodeText);
+        };
+
+        // handle arrow function shorthand
+        if (invalidAncestor.type === AST_NODE_TYPES.ArrowFunctionExpression) {
+          // handle wrapping with `void`
+          if (options.ignoreVoidOperator) {
+            return void context.report({
```

Comment:
nit - I personally don't like this code-style.
It looks like you're consuming the expression result so it returns `undefined`.
The style in use in this codebase is usually just this.

```suggestion
            return context.report({
```

Reply:
Reverted